DESIGN.md

```diff
@@ -0,0 +1,106 @@
+1. Resource control - since the user is able to limit the task resources we need to have a small layer for working with the file system.  
+
+### Resource control
+Three subsystems will be used for allowing a user to limit a job:
+1.  blkio - parameter _weight_ is used.
+1.  cpu - parameter _shares_ is used. 
+1. memory - parameter _limit_in_bytes_ is used. It sets the upper limit of memory available to a particular job. 
+
+Initially, we set up a **teleworker** group with blkio.weight and cpu.shares parameters set to 1000 (i.e. maximum).
+Every job launched without limiting any of the parameters will be placed directly to **cgroup.procs** of that group, so basically we will write the job PID into the following:
+```
+/sys/fs/cgroup/cpu/teleworker/cgroup.procs
+/sys/fs/cgroup/blkio/teleworker/cgroup.procs
+/sys/fs/cgroup/memory/teleworker/cgroup.procs
+```
+If a user provides any of the limits when creating a task, we create a new group inside of the _teleworker_ named by the ID of that task. For consistency we do that for all the resources (i.e. the group _cpu/teleworker/uuid_ will be created even if only memory and io limits were set by the user). This means that we write PID of any limited job to
+```
+/sys/fs/cgroup/<resource>/teleworker/<uuid>/cgroup.procs
+```
+
+When the job is finished or terminated we also remove the related directories.
+
+
+## Server
+
+Server will extensively use the library and will probably have only some basic auth logic (see [Security](#security)) and a layer of logic for storing the current system state in memory.  
+
+
+## Client
+
+Client is a simple CLI for the communication with the server. It will allow running a set of commands against the server:
+
+1. Start a job: passes a command provided by the user for direct execution on the server side. The ID of the job will be sent in response.
+If the user wants to limit resources that will be available to a job upon execution he is required to do it while sending a start request.
+No update functionality will be provided in terms of resources control - if the user wants to add or change some limits he is required to stop the job and create a new one with the required parameters.
+The following set of flags is used for this purpose, user can set the required limit in one of the groups:
+    * **mem** - memory limit for a job in megabytes
+    * **cpu** - cpu share in percents (1-100) available to this job
+    * **io** - proportion of I/O access (1-100) available to this job
+1. Stop the job: a user is required to provide a job ID for the job termination. The default behavior is to kill the task as it will trigger _SIGKILL_ to be sent for the command termination.
+1. Get the status of the job. Requires only the job ID to be sent, the user gets in return the job status, all the job resource limits set upon job creation and exit code (applies only if the job is in the finished or stopped status)
```

Comment:
Will a status request immediately after a stop request be guaranteed to show the command in the stopped state, or not?

Reply:
Now when the only option to stop command is by SIGKILL we can guarantee that, I've added it to the doc, thanks